concurrency/PendingReconstructionBlocks.java

```diff
@@ -0,0 +1,306 @@
+ *     that never made it.
+ *
+ ***************************************************/
+class PendingReconstructionBlocks {
+  private static final Logger LOG = BlockManager.LOG;
+
+  private final Map<BlockInfo, PendingBlockInfo> pendingReconstructions;
+  private final ArrayList<BlockInfo> timedOutItems;
+  Daemon timerThread = null;
+  private volatile boolean fsRunning = true;
+  private long timedOutCount = 0L;
+
+  //
+  // It might take anywhere between 5 to 10 minutes before
+  // a request is timed out.
+  //
+  private long timeout =
+      DFS_NAMENODE_RECONSTRUCTION_PENDING_TIMEOUT_SEC_DEFAULT * 1000;
+  private final static long DEFAULT_RECHECK_INTERVAL = 5 * 60 * 1000;
+
+  PendingReconstructionBlocks(long timeoutPeriod) {
+    if ( timeoutPeriod > 0 ) {
+      this.timeout = timeoutPeriod;
+    }
+    pendingReconstructions = new HashMap<>();
+    timedOutItems = new ArrayList<>();
+  }
+
+  void start() {
+    timerThread = new Daemon(new PendingReconstructionMonitor());
+    timerThread.start();
+  }
+
+  /**
+   * Add a block to the list of pending reconstructions
+   * @param block The corresponding block
+   * @param targets The DataNodes where replicas of the block should be placed
+   */
+  void increment(BlockInfo block, DatanodeDescriptor... targets) {
+    synchronized (pendingReconstructions) {
+      PendingBlockInfo found = pendingReconstructions.get(block);
+      if (found == null) {
+        pendingReconstructions.put(block, new PendingBlockInfo(targets));
+      } else {
+        found.incrementReplicas(targets);
+        found.setTimeStamp();
+      }
+    }
+  }
+
+  /**
+   * One reconstruction request for this block has finished.
+   * Decrement the number of pending reconstruction requests
+   * for this block.
+   *
+   * @param dn The DataNode that finishes the reconstruction
+   * @return true if the block is decremented to 0 and got removed.
+   */
+  boolean decrement(BlockInfo block, DatanodeDescriptor dn) {
+    boolean removed = false;
+    synchronized (pendingReconstructions) {
+      PendingBlockInfo found = pendingReconstructions.get(block);
+      if (found != null) {
+        LOG.debug("Removing pending reconstruction for {}", block);
+        found.decrementReplicas(dn);
+        if (found.getNumReplicas() <= 0) {
+          pendingReconstructions.remove(block);
+          removed = true;
+        }
+      }
+    }
+    return removed;
+  }
+
+  /**
+   * Remove the record about the given block from pending reconstructions.
+   *
+   * @param block
+   *          The given block whose pending reconstruction requests need to be
+   *          removed
+   */
+  void remove(BlockInfo block) {
+    synchronized (pendingReconstructions) {
+      pendingReconstructions.remove(block);
+    }
+  }
+
+  public void clear() {
+    synchronized (pendingReconstructions) {
+      pendingReconstructions.clear();
+      timedOutItems.clear();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **130**. At this line the following field is not protected by synchronization: `timedOutItems`. Your code typically protects this field with synchronization on object: `timedOutItems`.

**Fix**  
 Consider protecting the above field with synchronization on object: `timedOutItems`. For example, like at lines: **263** and **163**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **163**, **174**, **177**, **178**, **180**, and **263**.